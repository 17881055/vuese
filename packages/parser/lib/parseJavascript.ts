import traverse, { NodePath } from '@babel/traverse'
import generate from '@babel/generator'
import * as bt from '@babel/types'
import { getComments, CommentResult } from './jscomments'
import {
  PropsResult,
  PropType,
  ParserOptions,
  EventResult,
  MethodResult
} from './index'
import {
  getValueFromGenerate,
  isVueOption,
  runFunction,
  getArgumentFromPropDecorator,
  getEmitDecorator,
  getComponentDescribe
} from './helper'

export function parseJavascript(ast: bt.File, options: ParserOptions = {}) {
  const seenEvent = new Set()

  traverse(ast, {
    ExportDefaultDeclaration(rootPath: NodePath<bt.ExportDefaultDeclaration>) {
      // Get a description of the component
      if (options.onDesc) options.onDesc(getComponentDescribe(rootPath.node))

      rootPath.traverse({
        ObjectProperty(path: NodePath<bt.ObjectProperty>) {
          const { onProp, onMethod, onName } = options
          // Processing name
          if (isVueOption(path, 'name')) {
            let componentName = (path.node.value as bt.StringLiteral).value
            if (onName) onName(componentName)
          }

          // Processing props
          if (isVueOption(path, 'props')) {
            const valuePath = path.get('value')

            if (bt.isArrayExpression(valuePath.node)) {
              // An array of strings
              const propsValue: [] = getValueFromGenerate(valuePath.node)
              const propsRes: PropsResult[] = normalizeProps(propsValue)
              propsRes.forEach(prop => {
                if (onProp) onProp(prop)
              })
            } else if (bt.isObjectExpression(valuePath.node)) {
              // An object
              valuePath.traverse({
                ObjectProperty(propPath: NodePath<bt.ObjectProperty>) {
                  // Guarantee that this is the prop definition
                  if (propPath.parentPath === valuePath) {
                    const name = propPath.node.key.name
                    const propValueNode = propPath.node.value
                    const result: PropsResult = {
                      name,
                      type: null,
                      describe: getComments(propPath.node).default
                    }

                    processPropValue(propValueNode, result)

                    if (onProp) onProp(result)
                  }
                }
              })
            }
          }

          // Processing methods
          if (isVueOption(path, 'methods')) {
            const properties = (path.node
              .value as bt.ObjectExpression).properties.filter(
              n => bt.isObjectMethod(n) || bt.isObjectProperty(n)
            ) as (bt.ObjectMethod | bt.ObjectProperty)[]

            properties.forEach(node => {
              const commentsRes: CommentResult = getComments(node)
              // Collect only methods that have @vuese annotations
              if (commentsRes.vuese) {
                const result: MethodResult = {
                  name: node.key.name,
                  describe: commentsRes.default,
                  argumentsDesc: commentsRes.arg
                }
                if (onMethod) onMethod(result)
              }
            })
          }
        },
        CallExpression(path: NodePath<bt.CallExpression>) {
          const node = path.node
          // this.$emit()
          if (
            bt.isMemberExpression(node.callee) &&
            bt.isThisExpression(node.callee.object) &&
            bt.isIdentifier(node.callee.property) &&
            node.callee.property.name === '$emit' &&
            bt.isExpressionStatement(path.parentPath.node)
          ) {
            const { onEvent } = options
            const args = node.arguments
            const result: EventResult = {
              name: '',
              isSync: false,
              syncProp: ''
            }
            const firstArg = args[0]
            if (firstArg && bt.isStringLiteral(firstArg)) {
              processEventName(firstArg.value, result)
            }

            if (!result.name || seenEvent.has(result.name)) return
            seenEvent.add(result.name)

            const allComments: CommentResult = getComments(path.parentPath.node)
            result.describe = allComments.default
            result.argumentsDesc = allComments.arg

            if (onEvent) onEvent(result)
          }
        }
      })
    },
    // Class style component
    ClassProperty(path: NodePath<bt.ClassProperty>) {
      const propDecoratorArg = getArgumentFromPropDecorator(path.node)
      if (!propDecoratorArg) return

      const result: PropsResult = {
        name: (path.node.key as bt.Identifier).name,
        type: null,
        describe: getComments(path.node).default
      }
      processPropValue(propDecoratorArg, result)

      if (options.onProp) options.onProp(result)
    },
    ClassMethod(path: NodePath<bt.ClassMethod>) {
      const node = path.node
      const commentsRes: CommentResult = getComments(node)
      // Collect only methods that have @vuese annotations
      if (commentsRes.vuese) {
        const result: MethodResult = {
          name: (node.key as bt.Identifier).name,
          describe: commentsRes.default,
          argumentsDesc: commentsRes.arg
        }
        if (options.onMethod) options.onMethod(result)
      }

      // @Emit
      const emitDecorator = getEmitDecorator(node.decorators)
      if (emitDecorator) {
        const result: EventResult = {
          name: '',
          isSync: false,
          syncProp: ''
        }
        const args = (emitDecorator.expression as bt.CallExpression).arguments
        if (args && args.length && bt.isStringLiteral(args[0])) {
          result.name = (args[0] as bt.StringLiteral).value
        } else {
          if (bt.isIdentifier(node.key)) {
            result.name = node.key.name.replace(/([A-Z])/g, '-$1').toLowerCase()
          }
        }
        if (result.name) processEventName(result.name, result)

        if (!result.name || seenEvent.has(result.name)) return
        seenEvent.add(result.name)

        const allComments: CommentResult = getComments(node)
        result.describe = allComments.default
        result.argumentsDesc = allComments.arg

        if (options.onEvent) options.onEvent(result)
      }
    }
  })
}

function normalizeProps(props: string[]): PropsResult[] {
  return props.map(prop => ({
    type: null,
    name: prop
  }))
}

function getTypeByTypeNode(typeNode: bt.Node): PropType {
  if (bt.isIdentifier(typeNode)) return typeNode.name
  if (bt.isArrayExpression(typeNode)) {
    if (!typeNode.elements.length) return null

    return typeNode.elements
      .filter(node => node && bt.isIdentifier(node))
      .map(node => (node as bt.Identifier).name)
  }

  return null
}

// The `type` of a prop should be an array of constructors or constructors
// eg. String or [String, Number]
function isAllowPropsType(typeNode: bt.Node): boolean {
  return bt.isIdentifier(typeNode) || bt.isArrayExpression(typeNode)
}

function hasFunctionTypeDef(type: PropType): boolean {
  if (typeof type === 'string') {
    return type.toLowerCase() === 'function'
  } else if (Array.isArray(type)) {
    return type.map(a => a.toLowerCase()).some(b => b === 'function')
  }
  return false
}

function processEventName(eventName: string, result: EventResult) {
  const syncRE = /^update:(.+)/
  const eventNameMatchs = eventName.match(syncRE)
  result.name = eventName
  // Mark as .sync
  if (eventNameMatchs) {
    result.isSync = true
    result.syncProp = eventNameMatchs[1]
  }
}

function processPropValue(propValueNode: bt.Node, result: PropsResult) {
  if (isAllowPropsType(propValueNode)) {
    result.type = getTypeByTypeNode(propValueNode)
  } else if (bt.isObjectExpression(propValueNode)) {
    if (!propValueNode.properties.length) return

    const allPropNodes = propValueNode.properties

    const typeNode: any[] = allPropNodes.filter((node: any) => {
      if (node.key.name === 'type') {
        return true
      }
      return false
    })
    const otherNodes = allPropNodes.filter((node: any) => {
      if (node.key.name !== 'type') {
        return true
      }
      return false
    })

    // Prioritize `type` before processing `default`.
    // Because the difference in `type` will affect the way `default` is handled.
    if (typeNode.length > 0) {
      result.type = getTypeByTypeNode(typeNode[0].value)
      // Get descriptions of the type
      const typeDesc: string[] = getComments(typeNode[0]).default
      if (typeDesc.length > 0) {
        result.typeDesc = typeDesc
      }
    }

    otherNodes.forEach((node: any) => {
      const n = node.key.name
      if (n === 'default') {
        if (!hasFunctionTypeDef(result.type) && bt.isFunction(node.value)) {
          result.default = runFunction(node.value)
        } else {
          if (bt.isObjectMethod(node)) {
            result.default = generate(node).code
          } else {
            result.default = generate(node.value).code
          }
        }

        // Get descriptions of the default value
        const defaultDesc: string[] = getComments(node).default
        if (defaultDesc.length > 0) {
          result.defaultDesc = defaultDesc
        }
      } else if (n === 'required') {
        if (bt.isBooleanLiteral(node.value)) {
          result.required = node.value.value
        }
      } else if (n === 'validator') {
        if (bt.isObjectMethod(node)) {
          result.validator = generate(node).code
        } else {
          result.validator = generate(node.value).code
        }

        // Get descriptions of the validator
        const validatorDesc: string[] = getComments(node).default
        if (validatorDesc.length > 0) {
          result.validatorDesc = validatorDesc
        }
      }
    })
  }
}
