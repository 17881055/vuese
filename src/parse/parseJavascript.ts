import traverse from '@babel/traverse'
import generate from '@babel/generator'
import * as bt from '@babel/types'
import { getComments } from './jscomments'
import { PropsResult, PropType, ParserOptions } from './index'
import { getValueFromGenerate, isPropsOption, runFunction } from '../helpers'
import { isArray } from 'util'

const mainTraveres = {
  ObjectProperty(path: any) {
    // Processing props
    if (isPropsOption(path)) {
      const valuePath = path.get('value')
      const { onProp } = (this as any).options
      let res: PropsResult[] = []
      if (bt.isArrayExpression(valuePath.node)) {
        // An array of strings
        const propsValue: [] = getValueFromGenerate(valuePath.node)
        res = normalizeProps(propsValue)
      } else if (bt.isObjectExpression(valuePath.node)) {
        // An object
        valuePath.traverse({
          ObjectProperty(propPath: any) {
            // Guarantee that this is the prop definition
            if (propPath.parentPath === valuePath) {
              const name = propPath.node.key.name
              const vPath = propPath.get('value')
              const result: PropsResult = {
                name,
                type: null,
                describe: getComments(propPath)
              }

              if (isAllowPropsType(vPath.node)) {
                result.type = getTypeByPath(vPath)
              } else if (bt.isObjectExpression(vPath.node)) {
                if (!vPath.node.properties.length) return

                const allPropNodes = vPath.get('properties.0').container
                const typeNode = allPropNodes.filter((node: any, i: number) => {
                  if (node.key.name === 'type') {
                    node.$$selfPath = vPath.get(`properties.${i}`)
                    return true
                  }
                  return false
                })
                const otherNodes = allPropNodes.filter(
                  (node: any, i: number) => {
                    if (node.key.name !== 'type') {
                      node.$$selfPath = vPath.get(`properties.${i}`)
                      return true
                    }
                    return false
                  }
                )

                // Prioritize `type` before processing `default`.
                // Because the difference in `type` will affect the way `default` is handled.
                if (typeNode.length > 0) {
                  result.type = getTypeByPath(
                    typeNode[0].$$selfPath.get('value')
                  )
                  // Get descriptions of the type
                  const typeDesc: string[] = getComments(typeNode[0].$$selfPath)
                  if (typeDesc.length > 0) {
                    result.typeDesc = typeDesc
                  }
                }

                otherNodes.forEach((node: any) => {
                  const n = node.key.name
                  if (n === 'default') {
                    if (
                      !hasFunctionTypeDef(result.type) &&
                      bt.isFunction(node.value)
                    ) {
                      result.default = runFunction(node.value)
                    } else {
                      result.default = generate(node.value).code
                    }

                    // Get descriptions of the default value
                    const defaultDesc: string[] = getComments(node.$$selfPath)
                    if (defaultDesc.length > 0) {
                      result.defaultDesc = defaultDesc
                    }
                  } else if (n === 'required') {
                    if (bt.isBooleanLiteral(node.value)) {
                      result.required = node.value.value
                    }
                  } else if (n === 'validator') {
                    result.validator = generate(node.value).code

                    // Get descriptions of the validator
                    const validatorDesc: string[] = getComments(node.$$selfPath)
                    if (validatorDesc.length > 0) {
                      result.validatorDesc = validatorDesc
                    }
                  }
                })
              }
              res.push(result)
            }
          }
        })
      }
      if (onProp) onProp(res)
    }
  }
}

function normalizeProps(props: string[]): PropsResult[] {
  return props.map(prop => ({
    type: null,
    name: prop
  }))
}

function getTypeByPath(typePath: any): PropType {
  const typeNode = typePath.node
  if (bt.isIdentifier(typeNode)) return typeNode.name
  if (bt.isArrayExpression(typeNode)) {
    if (!typeNode.elements.length) return null
    return typePath
      .get('elements.0')
      .container.filter((node: any) => bt.isIdentifier(node))
      .map((node: any) => node.name)
  }
  return null
}

// The `type` of a prop should be an array of constructors or constructors
// eg. String or [String, Number]
function isAllowPropsType(typeNode: any): boolean {
  return bt.isIdentifier(typeNode) || bt.isArrayExpression(typeNode)
}

function hasFunctionTypeDef(type: PropType): boolean {
  if (typeof type === 'string') {
    return type.toLowerCase() === 'function'
  } else if (isArray(type)) {
    return type.map(a => a.toLowerCase()).some(b => b === 'function')
  }
  return false
}

export default function(ast: any, options: ParserOptions = {}) {
  traverse(ast, {
    ExportDefaultDeclaration(path: any) {
      path.traverse(mainTraveres, {
        options
      })
    }
  })
}
