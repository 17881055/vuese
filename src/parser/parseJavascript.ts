import traverse from '@babel/traverse'
import generate from '@babel/generator'
import * as bt from '@babel/types'
import { getComments, CommentResult } from './jscomments'
import {
  PropsResult,
  PropType,
  ParserOptions,
  EventResult,
  MethodResult
} from './index'
import { getValueFromGenerate, isVueOption, runFunction } from '../helpers'
import { isArray } from 'util'

const mainTraveres = {
  ObjectProperty(path: any) {
    const { onProp, onMethod, onName } = (this as any).options
    // Processing name
    if (isVueOption(path, 'name')) {
      if (onName) onName(path.node.value.value)
    }

    // Processing props
    if (isVueOption(path, 'props')) {
      const valuePath = path.get('value')
      let res: PropsResult[] = []
      if (bt.isArrayExpression(valuePath.node)) {
        // An array of strings
        const propsValue: [] = getValueFromGenerate(valuePath.node)
        res = normalizeProps(propsValue)
      } else if (bt.isObjectExpression(valuePath.node)) {
        // An object
        valuePath.traverse({
          ObjectProperty(propPath: any) {
            // Guarantee that this is the prop definition
            if (propPath.parentPath === valuePath) {
              const name = propPath.node.key.name
              const vPath = propPath.get('value')
              const result: PropsResult = {
                name,
                type: null,
                describe: getComments(propPath.node).default
              }

              if (isAllowPropsType(vPath.node)) {
                result.type = getTypeByPath(vPath)
              } else if (bt.isObjectExpression(vPath.node)) {
                if (!vPath.node.properties.length) return

                const allPropNodes = vPath.get('properties.0').container
                const typeNode = allPropNodes.filter((node: any, i: number) => {
                  if (node.key.name === 'type') {
                    node.$$selfPath = vPath.get(`properties.${i}`)
                    return true
                  }
                  return false
                })
                const otherNodes = allPropNodes.filter(
                  (node: any, i: number) => {
                    if (node.key.name !== 'type') {
                      node.$$selfPath = vPath.get(`properties.${i}`)
                      return true
                    }
                    return false
                  }
                )

                // Prioritize `type` before processing `default`.
                // Because the difference in `type` will affect the way `default` is handled.
                if (typeNode.length > 0) {
                  result.type = getTypeByPath(
                    typeNode[0].$$selfPath.get('value')
                  )
                  // Get descriptions of the type
                  const typeDesc: string[] = getComments(typeNode[0]).default
                  if (typeDesc.length > 0) {
                    result.typeDesc = typeDesc
                  }
                }

                otherNodes.forEach((node: any) => {
                  const n = node.key.name
                  if (n === 'default') {
                    if (
                      !hasFunctionTypeDef(result.type) &&
                      bt.isFunction(node.value)
                    ) {
                      result.default = runFunction(node.value)
                    } else {
                      if (bt.isObjectMethod(node)) {
                        result.default = generate(node).code
                      } else {
                        result.default = generate(node.value).code
                      }
                    }

                    // Get descriptions of the default value
                    const defaultDesc: string[] = getComments(node).default
                    if (defaultDesc.length > 0) {
                      result.defaultDesc = defaultDesc
                    }
                  } else if (n === 'required') {
                    if (bt.isBooleanLiteral(node.value)) {
                      result.required = node.value.value
                    }
                  } else if (n === 'validator') {
                    if (bt.isObjectMethod(node)) {
                      result.validator = generate(node).code
                    } else {
                      result.validator = generate(node.value).code
                    }

                    // Get descriptions of the validator
                    const validatorDesc: string[] = getComments(node).default
                    if (validatorDesc.length > 0) {
                      result.validatorDesc = validatorDesc
                    }
                  }
                })
              }
              res.push(result)
            }
          }
        })
      }
      if (onProp) onProp(res)
    }

    // Processing methods
    if (isVueOption(path, 'methods')) {
      const properties = path.node.value.properties
      properties.forEach((node: any) => {
        const commentsRes: CommentResult = getComments(node)
        // Collect only methods that have @vuese annotations
        if (commentsRes.vuese) {
          const result: MethodResult = {
            name: node.key.name,
            describe: commentsRes.default,
            argumentsDesc: commentsRes.arg
          }
          if (onMethod) onMethod(result)
        }
      })
    }
  },
  CallExpression(path: any) {
    const node = path.node
    // this.$emit()
    if (
      bt.isMemberExpression(node.callee) &&
      bt.isThisExpression(node.callee.object) &&
      bt.isIdentifier(node.callee.property) &&
      node.callee.property.name === '$emit' &&
      bt.isExpressionStatement(path.parentPath.node)
    ) {
      const { onEvent } = (this as any).options
      const args = node.arguments
      const result: EventResult = {
        name: ''
      }
      if (args.length && bt.isStringLiteral(args[0])) {
        result.name = args[0].value
      }

      if (!result.name || (this as any).seenEvent.has(result.name)) return
      ;(this as any).seenEvent.add(result.name)

      const allComments: CommentResult = getComments(path.parentPath.node)
      result.describe = allComments.default
      result.argumentsDesc = allComments.arg

      if (onEvent) onEvent(result)
    }
  }
}

function normalizeProps(props: string[]): PropsResult[] {
  return props.map(prop => ({
    type: null,
    name: prop
  }))
}

function getTypeByPath(typePath: any): PropType {
  const typeNode = typePath.node
  if (bt.isIdentifier(typeNode)) return typeNode.name
  if (bt.isArrayExpression(typeNode)) {
    if (!typeNode.elements.length) return null
    return typePath
      .get('elements.0')
      .container.filter((node: any) => bt.isIdentifier(node))
      .map((node: any) => node.name)
  }
  return null
}

// The `type` of a prop should be an array of constructors or constructors
// eg. String or [String, Number]
function isAllowPropsType(typeNode: any): boolean {
  return bt.isIdentifier(typeNode) || bt.isArrayExpression(typeNode)
}

function hasFunctionTypeDef(type: PropType): boolean {
  if (typeof type === 'string') {
    return type.toLowerCase() === 'function'
  } else if (isArray(type)) {
    return type.map(a => a.toLowerCase()).some(b => b === 'function')
  }
  return false
}

export default function(ast: any, options: ParserOptions = {}) {
  const seenEvent = new Set()
  traverse(ast, {
    ExportDefaultDeclaration(path: any) {
      path.traverse(mainTraveres, {
        options,
        seenEvent
      })
    }
  })
}
