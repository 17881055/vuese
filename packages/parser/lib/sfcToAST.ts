// Use vue-template-compiler/build to avoid detection of vue versions
// in vue-template-compiler/index.js
import { parseComponent, compile } from 'vue-template-compiler/build'
import { parse as babelParse, ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'

export interface AstResult {
  sourceType?: string
  jsAst?: bt.File
  templateAst?: object
}

export function sfcToAST(
  source: string,
  babelParserPlugins?: ParserPlugin[]
): AstResult {
  const sfc = parseComponent(source)
  const res: AstResult = {}
  if (sfc.script && sfc.script.content) {
    res.sourceType = sfc.script.lang || 'js'
    res.jsAst = babelParse(sfc.script.content, {
      sourceType: 'module',
      plugins: babelParserPlugins || [
        'objectRestSpread',
        'dynamicImport',
        'decorators-legacy',
        'classProperties',
        'typescript'
      ]
    })
  }
  if (sfc.template && sfc.template.content) {
    res.templateAst = compile(sfc.template.content, {
      comments: true
    }).ast
  }
  return res
}
