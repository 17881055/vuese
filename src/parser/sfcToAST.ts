import { parse as vueSFCParse } from '@vue/component-compiler-utils'
import * as vueTemplateCompiler from 'vue-template-compiler'
import { parse as babelParse, ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'

export interface AstResult {
  sourceType?: string
  jsAst?: bt.File
  templateAst?: object
}

export default function(
  source: string,
  babelParserPlugins?: ParserPlugin[]
): AstResult {
  const sfc = vueSFCParse({
    source: source,
    compiler: vueTemplateCompiler,
    needMap: false
  })
  const res: AstResult = {}
  if (sfc.script && sfc.script.content) {
    res.sourceType = sfc.script.lang || 'js'
    res.jsAst = babelParse(sfc.script.content, {
      sourceType: 'module',
      plugins: babelParserPlugins || [
        'objectRestSpread',
        'dynamicImport',
        'decorators-legacy',
        'typescript'
      ]
    })
  }
  if (sfc.template && sfc.template.content) {
    res.templateAst = vueTemplateCompiler.compile(sfc.template.content, {
      comments: true
    }).ast
  }
  return res
}
