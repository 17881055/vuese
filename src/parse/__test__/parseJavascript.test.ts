import parseJavascript from '../parseJavascript'
import { ParserOptions, PropsResult, EventResult } from '../index'
import * as path from 'path'
import * as fs from 'fs'
import sfcToAST, { AstResult } from '../sfcToAST'

function getAST(fileName: string): object {
  const p = path.resolve(__dirname, `../../../__fixtures__/${fileName}`)
  const source = fs.readFileSync(p, 'utf-8')
  return sfcToAST(source)
}

test('Ability to correctly handle props that is an array of string', () => {
  const sfc: AstResult = getAST('arrayProps.vue')
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      expect((propsRes as PropsResult[])[0]).toEqual({
        type: null,
        name: 'a'
      })
    }
  }
  parseJavascript(sfc.jsAst, options)
})

const sfc1: AstResult = getAST('objectProps.vue')
test('Is a prop using a shorthand type', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      expect((propsRes as PropsResult[])[0]).toEqual({
        name: 'a',
        type: 'String',
        describe: []
      })
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('`prop` defined using a type array', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      expect((propsRes as PropsResult[])[1]).toEqual({
        name: 'b',
        type: ['Number', 'String'],
        describe: []
      })
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('Execute the default function and get the default value correctly', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      const propRes = (propsRes as PropsResult[])[2]
      expect(propRes.name).toBe('c')
      expect(propRes.default).toEqual({
        val: 1
      })
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('Get the `required` value correctly', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      const propRes = (propsRes as PropsResult[])[2]
      expect(propRes.required).toBe(true)
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('The validator function should be used as a string representation', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      const propRes = (propsRes as PropsResult[])[2]
      expect(propRes.validator).toMatchSnapshot()
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('The `prop` that does not satisfy the `prop` writing specification should be treated as no type', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      expect((propsRes as PropsResult[])[3]).toEqual({
        name: 'd',
        type: null,
        describe: []
      })
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('When the `type` definition contains `Function`, should get a string representation of the `default` function.', () => {
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      const propRes = (propsRes as PropsResult[])[4]
      expect(propRes.name).toBe('e')
      expect(propRes.default).toMatchSnapshot()
    }
  }
  parseJavascript(sfc1.jsAst, options)
})

test('Get comments as a description', () => {
  const sfc: AstResult = getAST('commentProps.vue')
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      const propRes = (propsRes as PropsResult[])[0]
      expect((propRes.describe as []).length).toBe(3)
      expect(propRes.describe).toMatchSnapshot()
    }
  }
  parseJavascript(sfc.jsAst, options)
})

test('Gets a description of the default value and a description of the validator', () => {
  const sfc: AstResult = getAST('propFieldComment.vue')
  const options: ParserOptions = {
    onProp: (propsRes?: PropsResult[]) => {
      let propRes = (propsRes as PropsResult[])[0]
      expect((propRes.defaultDesc as string[]).length).toBe(1)
      expect(propRes.defaultDesc).toEqual(['An empty function'])

      propRes = (propsRes as PropsResult[])[1]
      expect((propRes.validatorDesc as string[]).length).toBe(1)
      expect(propRes.validatorDesc).toEqual(['Must be a number greater than 0'])

      propRes = (propsRes as PropsResult[])[2]
      expect((propRes.typeDesc as string[]).length).toBe(1)
      expect(propRes.typeDesc).toMatchSnapshot()
    }
  }
  parseJavascript(sfc.jsAst, options)
})

test('Correct handling of events', () => {
  const sfc: AstResult = getAST('emit.vue')
  const options: ParserOptions = {
    onEvent(eventRes?: EventResult) {
      expect((eventRes as EventResult).name).toBe('click')
      expect(((eventRes as EventResult).describe as string[]).length).toBe(1)
      expect(((eventRes as EventResult).argumentsDesc as string[]).length).toBe(
        1
      )
      expect((eventRes as EventResult).describe).toMatchSnapshot()
      expect((eventRes as EventResult).argumentsDesc).toMatchSnapshot()
    }
  }
  parseJavascript(sfc.jsAst, options)
})
