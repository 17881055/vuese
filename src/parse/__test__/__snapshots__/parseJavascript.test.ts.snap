// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correct handling of events 1`] = `
Array [
  "Triggered when clicked",
]
`;

exports[`Correct handling of events 2`] = `
Array [
  "The first parameter is a Boolean value that represents...",
]
`;

exports[`Get comments as a description 1`] = `
Array [
  "a string",
  "For input form",
  "and something",
]
`;

exports[`Gets a description of the default value and a description of the validator 1`] = `
Array [
  "'TOP' | 'LEFT'",
]
`;

exports[`The validator function should be used as a string representation 1`] = `
"function () {
  return true;
}"
`;

exports[`When the \`type\` definition contains \`Function\`, should get a string representation of the \`default\` function. 1`] = `"default() {}"`;
